File: js/systems/TargetLock.js
// js/systems/TargetLock.js
import { state } from '../state.js';
import { isInViewport } from './Raycast.js';

// Candidate pool for lock-on: planetoids (including permanent
// structures — the dome, maze entrance, etc.) and asteroids, plus
// goombas. SpaceGhosts (state.enemies) and fire bars (state.fireBars)
// are deliberately excluded for now — fire bars are fixed hazards, not
// really "targets," and enemies are left out per an explicit decision
// to keep the first version of this scoped down; easy to add either
// later if wanted. Whatever the player is CURRENTLY standing on is
// also excluded — locking onto the ground under his own feet doesn't
// make sense as a target.
function gatherLockCandidates() {
  const standingOn = (state.player && state.player.onSurface) ? state.player.currentPlanet : null;
  return [...state.planetoids, ...state.asteroids, ...state.goombas]
    .filter(c => c !== standingOn)
    .filter(isInViewport);
}

// Still valid to remain locked onto: the object hasn't been destroyed
// (still present in whichever array it came from), hasn't scrolled out
// of the viewport, and isn't the planet the player has since landed on
// (e.g. after a successful pull toward it — stepping onto something
// locked should release the lock, the same way it was never offered as
// a candidate in the first place while already standing on it). Checked
// every frame by Player.js's computeLeftArmAimAngle, which clears
// this.lockedTarget the moment this returns false, falling back to the
// normal aim priority chain.
export function isValidLockTarget(obj) {
  if (!obj) return false;
  const stillExists = state.planetoids.includes(obj) || state.asteroids.includes(obj) || state.goombas.includes(obj);
  if (!stillExists) return false;
  if (!isInViewport(obj)) return false;
  if (state.player && state.player.onSurface && state.player.currentPlanet === obj) return false;
  return true;
}

// Picks whichever candidate is the smallest angular step AWAY from
// referenceAngle in the given direction (+1 = clockwise, -1 =
// counter-clockwise) — shared by selectNextLockTarget and
// selectPreviousLockTarget below, which are otherwise identical except
// for which direction they sweep.
//
// Canvas/world space here is Y-DOWN, which flips the usual math
// convention: INCREASING an atan2 angle is a CLOCKWISE sweep on
// screen, not counterclockwise (visualize rotating from "pointing
// right," angle 0, toward "pointing right-and-down," a small positive
// angle — that reads as sweeping clockwise, same direction a clock's
// hands move from 3 toward 4-5 o'clock). So "next clockwise" is simply
// "smallest angle strictly greater than referenceAngle," and "next
// counter-clockwise" is the mirror of that — "smallest angle strictly
// LESS than referenceAngle" — both wrapping around the full circle
// when nothing qualifies directly on that side.
function pickNearestInDirection(candidates, playerPos, referenceAngle, direction) {
  let best = null;
  let bestDelta = Infinity;
  for (const c of candidates) {
    const angle = Math.atan2(c.pos.y - playerPos.y, c.pos.x - playerPos.x);
    // How far AWAY from referenceAngle, in the given direction, to
    // reach this candidate's angle, normalized to (0, 2π] — excludes
    // exactly 0 deliberately (a delta of 0 gets pushed all the way to
    // 2π instead), so a candidate sitting essentially at
    // referenceAngle itself is only ever picked as an actual last
    // resort, not preferentially re-selected over things that are
    // genuinely further along the sweep.
    let delta = ((angle - referenceAngle) * direction) % (Math.PI * 2);
    if (delta <= 0) delta += Math.PI * 2;
    if (delta < bestDelta) {
      bestDelta = delta;
      best = c;
    }
  }
  return best;
}

// currentTarget (if any) is excluded from the candidate pool in both
// functions below, so repeated presses actually advance through the
// group instead of being able to re-select the same object
// immediately; if it turns out to be the ONLY thing in the viewport,
// currentTarget is returned again rather than releasing the lock just
// because there's nothing else nearby.

// Next candidate CLOCKWISE from referenceAngle (R1) — i.e., whichever
// candidate the blaster would sweep across first if it kept rotating
// clockwise starting from referenceAngle.
export function selectNextLockTarget(currentTarget, referenceAngle, playerPos) {
  const candidates = gatherLockCandidates().filter(c => c !== currentTarget);
  if (candidates.length === 0) {
    return currentTarget || null;
  }
  return pickNearestInDirection(candidates, playerPos, referenceAngle, 1);
}

// Next candidate COUNTER-clockwise from referenceAngle (L1) — the
// mirror of selectNextLockTarget above. In the common case of a
// single R1 press followed by a single L1 press with nothing having
// changed in between, this naturally lands back on whatever was
// locked before the R1 press — R1's "nearest clockwise neighbor" and
// L1's "nearest counter-clockwise neighbor" are symmetric by
// construction, so it reads as a genuine undo, not just "some other
// direction."
export function selectPreviousLockTarget(currentTarget, referenceAngle, playerPos) {
  const candidates = gatherLockCandidates().filter(c => c !== currentTarget);
  if (candidates.length === 0) {
    return currentTarget || null;
  }
  return pickNearestInDirection(candidates, playerPos, referenceAngle, -1);
}

File: lua/systems/TargetLock.lua
-- lua/systems/TargetLock.lua
--
-- Port of js/systems/TargetLock.js.
-- Candidate-gathering/selection logic for the gamepad hard-lock (R1 =
-- next/clockwise, L1 = previous/counter-clockwise, Circle = release —
-- see GamePadInput.lua and Player:tryLockTargetNext/Previous/
-- clearLockTarget).

local state = require("lua.state")
local utils = require("lua.utils")

local TargetLock = {}

-- Candidate pool for lock-on: planetoids (including permanent
-- structures like the sky dome) and asteroids, plus ooombas.
-- isPullExempt objects (TiledTerrain's walkable shapes, JumpPlatform)
-- are excluded — they're terrain, not really "targets," the same
-- reasoning that already excludes them from pull-beam selection (see
-- Player:trySelectPullTarget). Whatever the player is CURRENTLY
-- standing on is also excluded — locking onto the ground under his
-- own feet doesn't make sense as a target. SpaceGhosts/fire bars are
-- deliberately left out too, matching the original's own scoped-down
-- first version.
local function gatherLockCandidates()
  local standingOn = (state.player and state.player.onSurface) and state.player.currentPlanet or nil
  local candidates = {}

  for _, p in ipairs(state.planetoids) do
    if p ~= standingOn and not p.isPullExempt
       and utils.isOnScreen(p.pos.x, p.pos.y, utils.boundingRadius(p), 0) then
      table.insert(candidates, p)
    end
  end

  if state.asteroids then
    for _, a in ipairs(state.asteroids) do
      if a ~= standingOn and utils.isOnScreen(a.pos.x, a.pos.y, utils.boundingRadius(a), 0) then
        table.insert(candidates, a)
      end
    end
  end

  if state.ooombas then
    for _, o in ipairs(state.ooombas) do
      if o ~= standingOn and utils.isOnScreen(o.pos.x, o.pos.y, utils.boundingRadius(o), 0) then
        table.insert(candidates, o)
      end
    end
  end

  return candidates
end

-- Still valid to remain locked onto: the object hasn't been destroyed
-- (still present in whichever array it came from), hasn't scrolled out
-- of the viewport, and isn't the planet the player has since landed on
-- (e.g. after a successful pull toward it). Checked every frame by
-- Player:computeLeftArmAimAngle, which clears lockedTarget the moment
-- this returns false, falling back to the normal aim priority chain.
function TargetLock.isValidLockTarget(obj)
  if not obj then return false end

  local stillExists = false
  for _, p in ipairs(state.planetoids) do
    if p == obj then stillExists = true; break end
  end
  if not stillExists and state.asteroids then
    for _, a in ipairs(state.asteroids) do
      if a == obj then stillExists = true; break end
    end
  end
  if not stillExists and state.ooombas then
    for _, o in ipairs(state.ooombas) do
      if o == obj then stillExists = true; break end
    end
  end
  if not stillExists then return false end

  if not utils.isOnScreen(obj.pos.x, obj.pos.y, utils.boundingRadius(obj), 0) then return false end
  if state.player and state.player.onSurface and state.player.currentPlanet == obj then return false end
  return true
end

-- Picks whichever candidate is the smallest angular step AWAY from
-- referenceAngle in the given direction (+1 = clockwise, -1 =
-- counter-clockwise) — shared by selectNextLockTarget and
-- selectPreviousLockTarget below, which are otherwise identical except
-- for which direction they sweep.
--
-- Canvas/world space here is Y-DOWN, same as the original, which flips
-- the usual math convention: INCREASING an atan2 angle is a CLOCKWISE
-- sweep on screen, not counter-clockwise. So "next clockwise" is
-- simply "smallest angle strictly greater than referenceAngle," and
-- "next counter-clockwise" is the mirror of that.
--
-- Lua's `%` is a floor-modulo (always same sign as the divisor, so the
-- raw result already lands in [0, 2*pi) regardless of the dividend's
-- sign) — unlike JS's `%`, which can come back negative and needs its
-- own explicit fixup for that. The `delta <= 0` fixup below is still
-- needed here, just for a narrower reason: pushing an exact-zero delta
-- (a candidate sitting essentially at referenceAngle itself) out to a
-- full 2*pi turn, so it's only ever picked as a genuine last resort.
local function pickNearestInDirection(candidates, playerX, playerY, referenceAngle, direction)
  local best, bestDelta = nil, math.huge
  for _, c in ipairs(candidates) do
    local angle = math.atan2(c.pos.y - playerY, c.pos.x - playerX)
    local delta = ((angle - referenceAngle) * direction) % (math.pi * 2)
    if delta <= 0 then delta = delta + math.pi * 2 end
    if delta < bestDelta then
      bestDelta = delta
      best = c
    end
  end
  return best
end

-- currentTarget (if any) is excluded from the candidate pool, so
-- repeated presses actually advance through the group instead of
-- re-selecting the same object immediately; if it turns out to be the
-- ONLY thing in the viewport, currentTarget is returned again rather
-- than releasing the lock just because there's nothing else nearby.
local function selectInDirection(currentTarget, referenceAngle, originPos, direction)
  local candidates = {}
  for _, c in ipairs(gatherLockCandidates()) do
    if c ~= currentTarget then table.insert(candidates, c) end
  end
  if #candidates == 0 then
    return currentTarget or nil
  end
  return pickNearestInDirection(candidates, originPos.x, originPos.y, referenceAngle, direction)
end

-- Next candidate CLOCKWISE from referenceAngle (R1).
function TargetLock.selectNextLockTarget(currentTarget, referenceAngle, originPos)
  return selectInDirection(currentTarget, referenceAngle, originPos, 1)
end

-- Next candidate COUNTER-clockwise from referenceAngle (L1) — the
-- mirror of selectNextLockTarget above. In the common case of a single
-- R1 press followed by a single L1 press with nothing having changed
-- in between, this naturally lands back on whatever was locked before
-- the R1 press, reading as a genuine undo.
function TargetLock.selectPreviousLockTarget(currentTarget, referenceAngle, originPos)
  return selectInDirection(currentTarget, referenceAngle, originPos, -1)
end

return TargetLock
